go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1402,32 +1391,66 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
 	// 2. FAILED_PRECONDITION errors. These are caused when a failover is in progress.
 	// If so, we don't want to suppress the error. This will allow VTGate to
 	// detect and perform buffering during failovers.
-	if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
-		sqlErr, ok := origErr.(*mysql.SQLError)
-		if ok {
-			sqlState := sqlErr.SQLState()
-			errnum := sqlErr.Number()
-			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s) during query: %s", errnum, sqlState, sqlparser.TruncateForLog(sql))
-			err = formatErrorWithCallerID(ctx, err)
+	var message string
+	sqlErr, ok := err.(*mysql.SQLError)
+	if ok {
+		sqlState := sqlErr.SQLState()
+		errnum := sqlErr.Number()
+		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s)%s: %s", errnum, sqlState, callerID, queryAsString(sql, nil))
+			if logMethod != nil {
+				message = fmt.Sprintf("%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, doubleTruncate(sql, bindVariables))
+			}
+		} else {
+			err = vterrors.Errorf(errCode, "%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, queryAsString(sql, bindVariables))
+			if logMethod != nil {
+				message = fmt.Sprintf("%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, doubleTruncate(sql, bindVariables))
+			}
+		}
+	} else {
+		err = vterrors.Errorf(errCode, "%v%s", err.Error(), callerID)
+		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+			if logMethod != nil {
+				message = fmt.Sprintf("%v: %v", err, doubleTruncate(sql, nil))
+			}
+		} else {
+			if logMethod != nil {
+				message = fmt.Sprintf("%v: %v", err, doubleTruncate(sql, bindVariables))
+			}
 		}
 	}
 
+	if logMethod != nil {
+		logMethod(message)
+	}
+
 	if logStats != nil {
 		logStats.Error = err
 	}
 
 	return err
 }
 
-func formatErrorWithCallerID(ctx context.Context, err error) error {
-	if err == nil {
-		return nil
+func doubleTruncate(sql string, bindVariables map[string]*querypb.BindVariable) string {
```

Comment:
I think renaming to `truncateSqlAndBindVars` would be clearer as would adding a comment.